float64 face_center_x
float64 face_center_y
float64 face_offset_x
float64 face_offset_y